src/core/lib/iomgr/executor/threadpool.h

```diff
@@ -0,0 +1,154 @@
+
+#include <grpc/grpc.h>
+
+#include "src/core/lib/gprpp/thd.h"
+#include "src/core/lib/iomgr/executor/mpmcqueue.h"
+
+namespace grpc_core {
+
+// A base abstract base class for threadpool.
+// Threadpool is an executor that maintains a pool of threads sitting around
+// and waiting for closures. A threadpool also maintains a queue of pending
+// closures, when closures appearing in the queue, the threads in pool will
+// pull them out and execute them.
+class ThreadPoolInterface {
+ public:
+  // Waits for all pending closures to complete, then shuts down thread pool.
+  virtual ~ThreadPoolInterface() {}
+
+  // Schedules a given closure for execution later.
+  // Depending on specific subclass implementation, this routine might cause
+  // current thread to be blocked (in case of unable to schedule).
+  // Closure should contain a function pointer and arguments it will take, more
+  // details for closure struct at /grpc/include/grpc/impl/codegen/grpc_types.h
+  virtual void Add(grpc_experimental_completion_queue_functor* closure)
+      GRPC_ABSTRACT;
+
+  // Returns the current number of pending closures
+  virtual int num_pending_closures() const GRPC_ABSTRACT;
+
+  // Returns the capacity of pool (number of worker threads in pool)
+  virtual int pool_capacity() const GRPC_ABSTRACT;
+
+  // Thread option accessor
+  virtual const Thread::Options& thread_options() const GRPC_ABSTRACT;
+
+  // Returns the thread name for threads in this ThreadPool.
+  virtual const char* thread_name() const GRPC_ABSTRACT;
+
+  GRPC_ABSTRACT_BASE_CLASS
+};
+
+// Worker thread for threadpool. Executes closures in the queue, until getting a
+// NULL closure.
+class ThreadPoolWorker {
+ public:
+  ThreadPoolWorker(const char* thd_name, ThreadPoolInterface* pool,
+                   MPMCQueueInterface* queue, Thread::Options& options,
+                   int index)
+      : queue_(queue), thd_name_(thd_name), index_(index) {
+    thd_ = Thread(thd_name,
+                  [](void* th) { static_cast<ThreadPoolWorker*>(th)->Run(); },
+                  this, nullptr, options);
+  }
+
+  ~ThreadPoolWorker() {}
+
+  void Start() { thd_.Start(); }
+  void Join() { thd_.Join(); }
+
+  // GRPC_ABSTRACT_BASE_CLASS
+
+ private:
+  // struct for tracking stats of thread
+  struct Stats {
+    gpr_timespec sleep_time;
+    Stats() { sleep_time = gpr_time_0(GPR_TIMESPAN); }
+  };
+
+  void Run();  // Pulls closures from queue and executes them
+
+  MPMCQueueInterface* queue_;  // Queue in thread pool to pull closures from
+  Thread thd_;                 // Thread wrapped in
+  Stats stats_;                // Stats to be collected in run time
+  const char* thd_name_;       // Name of thread
+  int index_;                  // Index in thread pool
+};
+
+// A fixed size thread pool implementation of abstract thread pool interface.
+// In this implementation, the number of threads in pool is fixed, but the
+// capacity of closure queue is unlimited.
```

Comment:
suggest another flexible size of the thread pool, balance the performance and time-delay of handling closures .

Reply:
We already have grpc_core::Executor https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/executor.h as a thread pool with dynamic size. We may add another implementation of ThreadPoolInterface with dynamic size in the future.